### PR TITLE
Move Native Messaging docs from apps to extensions.

### LINKED ISSUE #1053: No native messaging page for extensions
The native messaging API is available to extensions and apps, but trying to look at the reference for extensions [1] just redirects to apps [2].  Since apps pages are also branded with large deprecation warnings, this implies native messaging is deprecated.

We should have a native messaging page for extensions.

[1] https://developer.chrome.com/docs/extensions/nativeMessaging/ (or https://developer.chrome.com/docs/extensions/mv3/nativeMessaging/)
[2] https://developer.chrome.com/docs/apps/nativeMessaging/

### LINKED ISSUE #2146: _com.my_company.my_application_ instead of com.my_company.my_application
**Describe the bug**
On https://developer.chrome.com/docs/apps/nativeMessaging/ "_com.my_company.my_application_" is used through the article instead of correct "com.my_company.my_application".

I tried to follow the instructions and created file `/etc/chromium/native-messaging-hosts/_chromium_exec_.json` (this is actual literal name!) in my Linux (chromium-exec is name of my project). This didn't work. Then I tried `chromium_exec` instead of `_chromium_exec_` and it worked. So, your docs are wrong.

Even if you think that your wording is clear for readers, then at least use it consistently. Your article sometimes uses variant _com.my_company.my_application_ and sometimes com.my_company.my_application

### PR DESCRIPTION
**Pull request by oliverdunk**
Copies the nativeMessaging API docs over from the apps folder to the extensions folder and makes some small updates to the wording to accomodate this.

The goal of this is to get _something_ live, and not do too much more then checking the article for accuracy, so I'm hoping we can merge this fairly quickly.

Fixes #1053
Fixes #2146

### CODE REVIEW

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+- Invalid native messaging host name specified.
+  - Check whether the name contains any invalid characters. Only lowercase alphanumeric characters,
+    underscores and dots are allowed. A name cannot start or end with a dot, and a dot cannot be
+    followed by another dot.
+- Native host has exited.
+  - The pipe to the native messaging host was broken before the message was read by Chrome. This is most
+    likely initiated from your native messaging host.
+- Specified native messaging host not found.
+  - Is the name spelled correctly in the extension and in the manifest file?
+  - Is the manifest put in the right directory and with the correct name? See [native messaging host
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/nativeMessaging/index.md:164]:
```suggestion
  - Is the manifest in the right directory and with the correct name? See [native messaging host
```

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    location][native-messaging-host-location] for the expected formats.
+  - Is the manifest file in the correct format? In particular, is the JSON syntax correct and do the
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/nativeMessaging/index.md:166]:
```suggestion
  - Is the manifest file in the correct format? In particular, is the JSON valid and well-formed and do the
```

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  - Does the file specified in `path` exist? On Windows, paths may be relative, but on macOS and Linux,
+    the paths must be absolute.
+- Native messaging host _host name_ is not registered. (Windows-only)
+  - The native messaging host was not found in the Windows registry. Double-check using `regedit`
+    whether the key was really created and matches the required format as documented at [native
+    messaging host location][native-messaging-host-location].
+- Access to the specified native messaging host is forbidden.
+  - Is the extension's origin listed in `allowed_origins`?
+- Error when communicating with the native messaging host.
+  - This is a very common error and indicates an incorrect implementation of the communication protocol
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/nativeMessaging/index.md:177]: If this is such a common error, why not list it first? Then you don't need to say it's a common error. You probably don't need to regardless.

**Review by jpmedley — CHANGES_REQUESTED** [review]
Oliver,

I like the organization of this, and you're clear throughout. Most of my changes are formatting and style requests to be more consistent with the rest of dcc and our various style guides. I hope you don't fret about this. I'm the one paid to worry about these things.

Joe

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+The first argument to the native messaging host is the origin of the caller, usually
+`chrome-extension://[ID of allowed extension]`. This allows native messaging hosts to identify the
+source of the message when multiple extensions are specified in the `allowed_origins` key in the
+[native messaging host manifest][native-messaging-host-manifest].
+**_Warning_**: In Windows, in Chrome 54 and earlier, the origin was passed as the second parameter
+instead of the first parameter.
+
+When a messaging port is created using [runtime.connectNative][connect-native] Chrome starts native messaging
+host process and keeps it running until the port is destroyed. On the other hand, when a message is
+sent using [runtime.sendNativeMessage][send-native-message], without creating a messaging port, Chrome starts a new
```
[the remark this answers is not in the document]
Comment (tophf) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:96]: Not sure adding `()` is better, but if this is indeed important then it should be probably automated via a script because you're missing it on some methods even in this review and I see the same inconsistent use of () in the other articles like [messaging](https://developer.chrome.com/docs/extensions/mv3/messaging/).

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+@="C:\\path\\to\\nmh-manifest.json"
+```
+
+When Chrome looks for native messaging hosts, first the 32-bit registry is queried, then the 64-bit
+registry.
+
+On **macOS** and **Linux**, the location of the native messaging host's manifest file varies by the
+browser (Google Chrome or Chromium). The system-wide native messaging hosts are looked up at a fixed
+location, while the user-level native messaging hosts are looked up in a subdirectory within the
+[user profile directory][user-data-directory] called `NativeMessagingHosts`.
```
[the remark this answers is not in the document]
Comment (tophf) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:63]: Why? The path separator isn't a part of the name.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  - Chromium: `/etc/chromium/native-messaging-hosts/com.my_company.my_application.json`
+- Linux (user-specific, _default_ path)
+  - Google Chrome: `~/.config/google-chrome/NativeMessagingHosts/com.my_company.my_application.json`
+  - Chromium: `~/.config/chromium/NativeMessagingHosts/com.my_company.my_application.json`
+
+### Native messaging protocol {: #native-messaging-host-protocol }
+
+Chrome starts each native messaging host in a separate process and communicates with it using
+standard input (`stdin`) and standard output (`stdout`). The same format is used to send messages in
+both directions: each message is serialized using JSON, UTF-8 encoded and is preceded with 32-bit
```
[the remark this answers is not in the document]
Comment (tophf) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:82]: Now it's unclear where the format is defined. The colon was used because this was a "definition: explanation".

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+source of the message when multiple extensions are specified in the `allowed_origins` key in the
+[native messaging host manifest][native-messaging-host-manifest].
+**_Warning_**: In Windows, in Chrome 54 and earlier, the origin was passed as the second parameter
+instead of the first parameter.
+
+When a messaging port is created using [runtime.connectNative][connect-native] Chrome starts native messaging
+host process and keeps it running until the port is destroyed. On the other hand, when a message is
+sent using [runtime.sendNativeMessage][send-native-message], without creating a messaging port, Chrome starts a new
+native messaging host process for each message. In that case the first message generated by the host
+process is handled as a response to the original request, i.e. Chrome will pass it to the response
```
[the remark this answers is not in the document]
Comment (tophf) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:98]: This was a direct explanation of "handled as a response", which is an ambiguous concept without its own specific meaning, so it might be better to leave the explanation in the same sentence. In case you're sure starting a sentence reads better consider starting it with `That is` instead of a somewhat unusually placed `I.e.`.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+The location of the manifest file depends on the platform.
+
+On **Windows**, the manifest file can be located anywhere in the file system. The application
+installer must create registry key
+`HKEY_LOCAL_MACHINE\SOFTWARE\Google\Chrome\NativeMessagingHosts\com.my_company.my_application` or
+`HKEY_CURRENT_USER\SOFTWARE\Google\Chrome\NativeMessagingHosts\com.my_company.my_application`, and
+set default value of that key to the full path to the manifest file. For example, using the
+following command:
+
+```text
```
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:47]: I'm reluctant to set bash here since this is a windows server command, not bash, and the highlighting might not be correct. Text is nice because it just renders plainly.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+REG ADD "HKCU\Software\Google\Chrome\NativeMessagingHosts\com.my_company.my_application" /ve /t REG_SZ /d "C:\path\to\nmh-manifest.json" /f
+```
+
+or using the following `.reg` file:
+
+```text
```
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:53]: Same as above.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md]: I've rephrased this a little to hopefully address both comments.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+### Native messaging protocol {: #native-messaging-host-protocol }
+
+Chrome starts each native messaging host in a separate process and communicates with it using
+standard input (`stdin`) and standard output (`stdout`). The same format is used to send messages in
+both directions: each message is serialized using JSON, UTF-8 encoded and is preceded with 32-bit
+message length in native byte order. The maximum size of a single message from the native messaging
+host is 1 MB, mainly to protect Chrome from misbehaving native applications. The maximum size of the
+message sent to the native messaging host is 4 GB.
+
+The first argument to the native messaging host is the origin of the caller, usually
```
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:89]: There was already a paragraph on this so I've just moved it up.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md]: I've reworded to just avoid this entirely.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+  function (response) {
+    console.log('Received ' + response);
+  }
+);
+```
+
+## Debugging native messaging {: #native-messaging-debugging }
+
+When the native messaging host fails to start, writes to `stderr` or when it violates the
+communication protocol, output is written to the error log of Chrome. On Linux and macOS, this log
```
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md]: I've reworded this to try and make it clearer. In the future, it'd be helpful if you could be more specific about what threw you off so I can make sure to address that :)

File: site/en/docs/extensions/mv3/nativeMessaging/index.md (hunk 2 above)
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md]: I actually think this makes it more wordy and hard to follow.

**Review by oliverdunk — COMMENTED** [review]
@jpmedley I think there was some confusion here - I didn't write this article, and am simply copying it from the `apps` folder over to `extensions`.

Thanks for the feedback regardless. I've done my best to do a pass addressing some of your comments but avoided anything that would make this in to too much of a bigger piece of work.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
... and dots. The name cannot start or end with a dot, and a dot cannot be followed by another dot.</td></tr><tr><td><code>description</code></td><td>Short application description.</td></tr><tr><td><code>path</code></td><td>Path to the native messaging host binary. On Linux and macOS the path must be absolute. On Windows it can be relative to the directory in which the manifest file is located. The host process is started with the current directory set to the directory that contains the host binary. For example if this parameter is set to <code>C:\Application\nm_host.exe</code> then it will be started with current directory <code>C:\Application\</code>.</td></tr><tr><td><code>type</code></td><td>Type of the interface used to communicate with the native messaging host. Currently there is only one possible value for this parameter: <code>stdio</code>. It indicates that Chrome should use <code>stdin</code> and <code>stdout</code> to communicate with the host.</td></tr><tr><td><code>allowed_origins</code></td><td>List of extensions that should have access to the native messaging host. Wildcards such as <code>chrome-extension://*/*</code> are <em>not</em> allowed.</td></tr></tbody></table>
```
[the remark this answers is not in the document]
Comment (jpmedley) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md:34]: At the very least, the HTML needs to be formatted to make reviewing and editing easier.

What is it about the DL formatting that you think isn't clean.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md (hunk 2 above)
Comment (jpmedley) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md]: Isn't 'valid and well-formed' the more technically correct term though?

**Comment by jpmedley** [comment]
>I didn't write this article, and am simply copying it from the apps folder over to extensions.

I thought it might be and didn't think to look in the apps folder for the original. I still think it's a good idea to give it a good review before moving. So I appreciate your time.

File: site/en/docs/extensions/mv3/nativeMessaging/index.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (oliverdunk) [reply site/en/docs/extensions/mv3/nativeMessaging/index.md]: I kept the table but moved to MD formatting. Let me know if that looks good.

**Review by jpmedley — APPROVED** [review]
Thank you for updating the table. Please commit the formatting corrections I made to it before merging this.
